### ISSUE TITLE
Loop with final state checker

### ISSUE THREAD

**Issue by pwielgolaski** [gatling/gatling#3281]
I have requirement to query server for few times until some condition is not met.
The condition is final either FAIL or OK, where intermediate state could be like PENDING/RUNNING

I used asLongAs but I would like to fail test if final state is as FAIL.

I wrote simple test that show what I want to achieve and fail.

```scala
package io.gatling.test

import io.gatling.commons.stats.KO
import io.gatling.core.Predef._
import io.gatling.core.session.Expression
import io.gatling.core.structure.ChainBuilder
import io.gatling.http.Predef._

class MySimulation extends Simulation {

  val httpConf = http.baseURL("http://httpbin.org")

  val incIndex: Expression[Session] = { s =>
    val oldValue = s("index").as[Int]
    s.set("index", oldValue + 1)
  }

  val checkState: Expression[Boolean] = { s =>
    val page = s.get("page").asOption[String].map(_.toInt).getOrElse(0)
    page < 5
  }

  def repeatUntilWithCheck(name: Expression[String], condition: Expression[Boolean], statusCheck: Expression[Boolean])(chain: ChainBuilder) = {
    group(name)(asLongAs(condition)(chain)
      // want to decide if group is completed with success
      // below does not work - no change to state
      .exec(s => s.markAsFailed)
      // logGroupRequest is only available for private[gatling]
      // it make group KO, but no way to make assertion - all request are OK
      .exec(s => s.logGroupRequest(0, KO))
    )
  }

  val scn = scenario("BasicSimulation")
    .exec(s => s.set("index", 1))
    .exec(repeatUntilWithCheck("repeat until", checkState, statusCheck = false)(
      exec(http("try").get("/get?page=${index}").check(jsonPath("$.args.page").find.saveAs("page"))).exec(incIndex)))

  setUp(
    scn.inject(atOnceUsers(1))
  ).protocols(httpConf)

}

``` 

The core logic is in repeatUntilWithCheck, after few check I would like to use some function to make decision if "group" is KO/OK. 
It can be done with private API, but it is not perfect either.

I would love to make it part of core gatling steps, but if not at least possible to implement similar to what I did.
How could I wrote my own step that is combination of predefined steps.

**Comment by slandelle** [comment]
> // below does not work - no change to state
    .exec(s => s.markAsFailed)

It should fail the group.

> // it make group KO, but no way to make assertion - all request are OK

Indeed, assertions currently target requests no groups.
If you want to make a request fail, write a [custom Validator](https://github.com/gatling/gatling/blob/master/gatling-core/src/main/scala/io/gatling/core/check/CheckBuilder.scala#L162-L166).

**Comment by pwielgolaski** [comment]
> // below does not work - no change to state
> .exec(s => s.markAsFailed)
> It should fail the group.

I agree that it should, but it does not as it does not touch "blockStack" different than `TryMaxBlock`, which in my case is `GroupBlock`

I will take a look at custom Validator and check if it would be good for this case.

**Comment by slandelle** [comment]
> I agree that it should, but it does not as it does not touch "blockStack" different than TryMaxBlock, which in my case is GroupBlock

Fixed as of #3197

**Comment by pwielgolaski** [comment]
@slandelle Great to hear that it was fixed, but it does not end up with 2.2.5
Does it mean that it will be 3.x? Any timeline?

Also if you could point me a bit with "custom validator" that you mentioned.
I was hoping that it could be done on assertion level, but seems that assertion are not very extendable.

**Comment by slandelle** [comment]
> Does it mean that it will be 3.x?

Yes. New features we develop don't go in bug fixing releases.

> Any timeline?

No, not atm.

> Also if you could point me a bit with "custom validator" that you mentioned.

Please check documentation: http://gatling.io/docs/current/http/http_check/#http-check-validating
